odl/solvers/functional/derivatives.py

```diff
@@ -0,0 +1,326 @@
+
+        if not isinstance(functional.domain, FnBase):
+            raise TypeError('`functional.domain` has to be a `FnBase` '
+                            'instance')
+
+        self.functional = functional
+        if step is None:
+            # Use half of the number of digits as machine epsilon, this
+            # "usually" gives a good balance between precision and numerical
+            # stability.
+            self.step = np.finfo(functional.domain.dtype).eps ** (1 / 2)
+        else:
+            self.step = float(step)
+
+        self.method = str(method).lower()
+        if method not in ('backward', 'forward', 'central'):
+            raise ValueError('`method` not understood')
+
+        Operator.__init__(self, functional.domain, functional.domain,
+                          linear=functional.is_linear)
+
+    def _call(self, x):
+        """Return ``self(x)``."""
+        # The algorithm takes finite differences in one dimension at a time
+        # reusing the dx vector to improve efficiency.
+        dfdx = self.domain.zero()
+        dx = self.domain.zero()
+
+        if self.method == 'backward':
+            fx = self.functional(x)
+            for i in range(self.domain.size):
+                dx[i - 1] = 0
+                dx[i] = self.step
+                dfdx[i] = fx - self.functional(x - dx)
+        elif self.method == 'forward':
+            fx = self.functional(x)
+            for i in range(self.domain.size):
+                dx[i - 1] = 0  # reset step from last iteration
+                dx[i] = self.step
+                dfdx[i] = self.functional(x + dx) - fx
+        elif self.method == 'central':
+            for i in range(self.domain.size):
+                dx[i - 1] = 0  # reset step from last iteration
+                dx[i] = self.step / 2
+                dfdx[i] = self.functional(x + dx) - self.functional(x - dx)
+        else:
+            raise RuntimeError('unknown method')
+
+        dfdx /= self.step
+        return dfdx
+
+    def derivative(self, point):
+        """Return the derivative in point.
+
+        The derivative of the gradient is often called the hessian.
+
+        Parameters
+        ----------
+        point : `domain` `element-like`
+            The point that the derivative should be taken in.
+
+        Returns
+        -------
+        derivative : `NumericalDerivative`
+            Numerical estimate of the derivative. Uses the same method as this
+            operator does, but with half the number of significant digits in
+            the step size in order to preserve numerical stability.
+
+        Examples
+        --------
+        Compute a numerical estimate of the derivative of the L2^2 norm:
+
+        >>> import odl
+        >>> space = odl.rn(3)
+        >>> func = odl.solvers.L2NormSquared(space)
+        >>> grad = NumericalGradient(func)
+        >>> hess = grad.derivative([1, 1, 1])
+        >>> hess([1, 0, 0])
```

Comment:
This example is not correct in the way I described above. The Hessian is usually the bilinear functional that appears as the second term in the Taylor expansion of an operator (or function) around a fixed point. Most of the time, this functional is identified with its operator representation, which in turn is identified with its matrix representation (for a finite-dimensional space) in the sense that
```
H(p)(x, y) = <H_op(p) x, y> = x_vec^T H_mat(p) y_vec
```
The relation to the gradient is as follows: if `S` is our functional, then its derivative is represented by its gradient via
```
S'(p) x = <grad S(p), x>
```
and in the Taylor expansion one has
```
S(p + h * x) ~ S(p) + h S'(p) x + o(h) = S(p) + h <grad S(p), x> + o(h)
```
Now, the second derivative of `S` in `p` is the derivative of `S'(p)` *in the same point `p`*, in order to make sense in the Taylor formula:
```
S(p + h * x) ~ S(p) + h S'(p) x + h^2 S''(p)(x, x) + o(h^2)
```
(Bilinear operators are uniquely determined by their diagonal values.)
How does that relate to the gradient? Let's differentiate it: for all `y`, it is
```
[S'(p + h * x)y - S'(p)y] / h --> S''(p)(x, y)
```
which can be translated to gradients:
```
<grad S(p + h * x) - grad S(p), y> / h --> <(grad S)'(p) x, y>
```
This shows that the derivative of the gradient of `S` in `p` is the same as the bilinear functional `S''(p)` in the sense of identification with the operator in an operator-weighted inner product. However, this only holds if we choose the same point `p` in the last step!

When I think about it, this would be a good fit for the math documentation, at the very least it would be useful for us to recall the proper settings of gradients and derivatives of functionals in Hilbert spaces.